src/utils/TextPrinter.js

```diff
@@ -0,0 +1,207 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+/* @flow */
+
+import type { Descriptor, PropertyBinding } from "../types.js";
+import { EnvironmentRecord, type Binding } from "../environment.js";
+import { PrimitiveValue, AbstractValue, ObjectValue, Value } from "../values/index.js";
+import invariant from "../invariant.js";
+import {
+  Printer,
+  Generator,
+  type OperationDescriptorType,
+  type CustomGeneratorEntryType,
+  type OperationDescriptorData,
+} from "./generator.js";
+
+export class TextPrinter implements Printer {
+  constructor(printLine: string => void) {
+    this._printLine = printLine;
+    this._indent = "";
+    this._abstractValueIds = new Map();
+  }
+
+  _printLine: string => void;
+  _indent: string;
+  _abstractValueIds: Map<AbstractValue, number>;
+
+  _nest(): void {
+    this._indent += " ";
+  }
+  _unnest(): void {
+    this._indent = this._indent.substring(0, this._indent.length - 1);
+  }
+
+  _print(text: string): void {
+    this._printLine(this._indent + text);
+  }
+
+  printGeneratorEntry(
+    declared: void | AbstractValue | ObjectValue,
+    type: OperationDescriptorType | CustomGeneratorEntryType,
+    args: Array<Value>,
+    data: OperationDescriptorData,
+    metadata: { isPure: boolean, mutatesOnly: void | Array<Value> }
+  ): void {
+    let text;
+    if (declared !== undefined) {
+      invariant(declared.intrinsicName !== undefined);
+      text = `${declared.intrinsicName} := `;
+    } else {
+      text = "";
+    }
+    text += type;
+
+    let dataTexts = [];
+    if (args.length > 0) dataTexts.push(`args ${this.describeValues(args)}`);
+    if (data.unaryOperator !== undefined) dataTexts.push(data.unaryOperator); // used by UNARY_EXPRESSION
+    if (data.binaryOperator !== undefined) dataTexts.push(data.binaryOperator); // used by BINARY_EXPRESSION
```

Comment:
Presumably these are mutually exclusive and this if can be inside an else.

Reply:
Lots of invariants over this `data` structure could be enforced somewhere and then assumed here. But that's for another PR...